### ISSUE TITLE
useDefineForClassFields and parameter properties: missing error or bad emit

### ISSUE THREAD

**Issue by ajafff** [microsoft/TypeScript#36410]
**TypeScript Version:**  3.8.0-dev.20200123

**Search Terms:** useDefineForClassFields parameter

**Code**

```ts
// @useDefineForClassFields: true
// @target: esnext
class C {
  bar = this.foo.bar;
  constructor(private foo: {bar: string}) {}
}
```

**Expected behavior:**

Transpiled code works (`bar` is defined in the constructor as it's done for the downlevel emit) OR used-before-assignment error on `this.foo` in the initializer of `bar`.

**Actual behavior:**

Since the initializer of `bar` is run before the assignment of `foo`, there's a runtime error.
This code used to work for ages (of course without native class fields). A lot of code that uses dependency injection looks like this.

**Playground Link:** https://www.typescriptlang.org/play/?useDefineForClassFields=true&target=99&ts=3.8.0-dev.20200123&ssl=4&ssc=2&pln=1&pc=1#code/MYGwhgzhAEDC0G8BQ1oCMwCdoF5oBcALASwgDoAzAeyrI0wG4VpgqA7CfTAV2HyswAKAA6ZiANzD4AptGpUAXInpLOYtgHMAvgEpEWpFqA

**Related Issues:** 
I thought this was already discussed somewhere with @RyanCavanaugh and @sandersn, but I couldn't find an issue for that.
Found while looking at #36405. That PR would need an exception if parameter properties are involved IF you want to continue supporting the example code.

**Comment by sandersn** [comment]
Support for [[Define]] so far has given an error whenever code that used to work for [[Set]] is no longer valid. I think that an error with useDefineForClassFields: true is the correct thing here. This error should move to useDefineForClassFields: false at the same time as the other errors from #33401 and #33423.

**Comment by sandersn** [comment]
The emitted code is also wrong, but in a different way, for targets lower than esnext:

```js
// @useDefineForClassFields: true
// @target: es2019
class C {
    constructor(foo) {
        this.foo = foo;
        Object.defineProperty(this, "foo", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "bar", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.foo.bar
        });
    }
}
```

`"foo"`'s defineProperty should come before `this.foo = foo` (or should replace it entirely).

**Comment by sandersn** [comment]
Never mind, the bug I saw must be fixed because it only [repros in the playground](https://www.typescriptlang.org/play/index.html?useDefineForClassFields=true&target=6#code/PTAEAEFcGcFMBFYDMCWA7WAxA9gJwMIA2AhtNJirIQCbQBcoALrpLALABQIEjxuA5rEYNY0DAA9GnAMYkyofKADenUKABGfUAF4mACxTQAdEmzYjm3AG5VoadjTRmkaYzwAKAA64UAN2KMsKCm2AxKlgxOPmj8AL4AlMqxnLFAA).

**Comment by ajafff** [comment]
@sandersn yeah, you fixed that a while back in #34987
... and in that PR we already talked about the order of parameter properties vs. class field initializers.
